backend/api/pipeline.proto

```diff
@@ -151,14 +175,56 @@ message Pipeline {
   string description = 4;
 
   // Output. The input parameters for this pipeline.
+  // TODO(jingzhang36): replace this parameters field with the parameters field
+  // inside PipelineVersion when all usage of the former has been changed to use
+  // the latter.
   repeated Parameter parameters = 5;
 
   // The URL to the source of the pipeline. This is required when creating the
   // pipeine through CreatePipeline API.
+  // TODO(jingzhang36): replace this url field with the code_source field inside
+  // PipelineVersion when all usage of the former has been changed to use the
+  // latter.
   Url url = 7;
 
   // In case any error happens retrieving a pipeline field, only pipeline ID
   // and the error message is returned. Client has the flexibility of choosing
   // how to handle error. This is especially useful during listing call.
   string error = 6;
+
+  // Output only. The default version of the pipeline. As of now, the latest
+  // version is used as default. (In the future, if desired by customers, we
+  // can allow them to set default version.)
+  PipelineVersion default_version = 8;
 }
+
+message PipelineVersion {
+  // Output. Unique version ID. Generated by API server.
+  string id = 1;
+
+  // Optional input field. Version name provided by user.
+  string name = 2;
+
+  // Output. The time this pipeline version is created.
+  google.protobuf.Timestamp created_at = 3;
+
+  // Output. The input parameters for this pipeline.
+  repeated Parameter parameters = 4;
+
+  // Input.
+  oneof code_source {
```

Comment:
should they be modeled as oneof?
the code source is where the source code is located. the url is where the compiled package stored. they are not mutually exclusive.

Reply:
I feel the two approaches have their own pros and cons. If we have both source and package as optional fields, it is actually easier to code and allow users to specify both when both are available. The downside of this is when the two are inconsistent with each other (and we won't validate the provided source and package when those get set, since the validation work is not so trivial and probably not worthwhile). 

That being said, I would be happy to switch to the having-both-as-optional approach, since the code would be easier with that.....

Reply:
Also, if we allow both code source and package source to be present at the same time. Should we also allow multiple code sources? So we don't need oneof in CodeSource type.